Engine/Engine.cs

```diff
@@ -245,11 +245,15 @@ public void Run(AlgorithmNodePacket job, string assemblyPath)
                         // Algorithm runtime error:
                         if (algorithm.RunTimeError != null)
                         {
-                            throw algorithm.RunTimeError;
+                            throw new Exception("Algorithm.RunTimeError", algorithm.RunTimeError); //Keep original exception as inner exception to preserve stack trace
                         }
                     }
                     catch (Exception err)
                     {
+                        while (err != null && err.InnerException != null)
+                        {
+                            err = err.InnerException; //Find inner most exception as this will be most useful to algorithm designer
```

Comment:
I agree the exception should be thrown as @robcodes did, to avoid losing the inner exception (in the stack trace it will not be included though, but at least the information is in `err.InnerException`).
To recursively include **all** the inner exceptions in the logging, `err.ToString()` should be used instead of `err.StackTrace`, replacing:
`_algorithmHandlers.Results.RuntimeError(message, err.StackTrace);`
with
`_algorithmHandlers.Results.RuntimeError(message, err.ToString());`

The line below should probably be changed as well:
`_systemHandlers.Api.SetAlgorithmStatus(job.AlgorithmId, AlgorithmStatus.RuntimeError, message + " Stack Trace: " + err.StackTrace);`